bitwarden_access_manager.py

```diff
@@ -0,0 +1,672 @@
+        result = (
+            "[[collection]]\n"
+            f"collection_id = {self.id}\n"
+            # Splicing the string is safe here, because Bitwarden repo names are
+            # very restrictive and do not contain quotes.
+            f'external_id = "{self.external_id}"\n'
+        )
+
+        if self.member_access is not None:
+            member_access_lines = ["  " + a.format_toml() for a in sorted(self.member_access)]
+            if len(member_access_lines) > 0:
+                result = (
+                    result + "member_access = [\n" + ",\n".join(member_access_lines) + ",\n]\n"
+                )
+            else:
+                result = result + "member_access = []\n"
+
+        if self.group_access is not None:
+            group_access_lines = ["  " + a.format_toml() for a in sorted(self.group_access)]
+            if len(group_access_lines) > 0:
+                result = (
+                    result + "group_access = [\n" + ",\n".join(group_access_lines) + ",\n]"
+                )
+            else:
+                result = result + "group_access = []"
+        return result
+
+class BitwardenClient(NamedTuple):
+    connection: HTTPSConnection
+    bearer_token: str
+
+    @staticmethod
+    def new(client_id: str, client_secret: str) -> BitwardenClient:
+        response = requests.post(
+            "https://identity.bitwarden.com/connect/token",
+            headers={ "Content-Type": "application/x-www-form-urlencoded" },
+            data={
+                "grant_type": "client_credentials",
+                "scope": "api.organization",
+                "Accept": "application/json",
+            },
+            auth=(client_id, client_secret),
+        )
+        bearer_token = response.json()["access_token"]
+        return BitwardenClient(HTTPSConnection("api.bitwarden.com"), bearer_token)
+
+    def _http_get(self, url: str) -> HTTPResponse:
+        self.connection.request(
+            method="GET",
+            url=url,
+            headers={
+                "Accept": "application/json",
+                "Authorization": f"Bearer {self.bearer_token}",
+            },
+        )
+
+        return self.connection.getresponse()
+
+    def get_group(self, id: str) -> Any:
+        return json.load(self._http_get(f"/public/groups/{id}"))
+
+    def get_groups(self) -> Iterable[Group]:
+        data = self._http_get(f"/public/groups")
+        groups = json.load(data)
+        for group in groups["data"]:
+            yield Group(
+                id=group["id"],
+                name=group["name"],
+                access_all=group["accessAll"],
+            )
+
+    def get_collection_members(self, groups: Any) -> Iterable[MemberCollectionAccess]:
+            for group in groups:
+                group_id = group["id"]
+
+                data = self._http_get(f"/public/groups/{group_id}/member-ids")
+                member_ids = json.load(data)
+
+                for member_id in member_ids:
```

Comment:
Now we need to make a lot of API calls, and many of them might be duplicated.

We need to load the organization members anyway to check those against the toml file, so we could take a dictionary that maps member id to `Member` as input to this function, and then it can just look up in the dict instead of making an API call.

Reply:
I implemented in [e3a0b84095](https://github.com/ChorusOne/github-access-manager/pull/10/commits/e3a0b84095be070ebdf48f70ec3423828c62f74a)